scripts/service.py

```diff
@@ -0,0 +1,286 @@
+            self.fifo = os.open(self.pipePath, os.O_RDWR | os.O_NONBLOCK)
+            self.epollObj.register(self.fifo, select.EPOLLIN)
+        except OSError as err:
+            self.logErr("open file read error:" + str(err))
+            raise
+
+    def _executeSub(self):
+        """ execute the passed script with subprocess """
+        self.logInfo("will invoke scripts:" + self.script)
+        try:
+            subprocess.Popen(self.script)
+        except OSError as err:
+            self.logErr("execute script error:" + str(err))
+            
+    def _readRequest(self):
+        """ read all that is written in the pipe, so in case of input storm,
+            just do work once.
+        """
+        if self.fifo != None:
+            try:
+                while True:
+                    self.buffer = os.read(self.fifo, BUFFER_SIZE)
+                    if len(self.buffer) == 0:
+                        # hung up
+                        self.logErr("hung up found while read, exiting")
+                        raise IOError(errno.EPIPE, 'hung up occurred while read pipe', self.pipePath)
+            except OSError as err:
+                if err.errno == errno.EAGAIN or err.errno == errno.EWOULDBLOCK:
+                    return True # assume already read as invoked after poll
+                else:
+                    self.logErr("Error occurred while read pipe:" + str(err))
+                    raise
+        else:
+            return False
+            
+    def _waitForInput(self):
+        """ hang on the poll to wait for incoming requests.
+            As we hold one writer, there should be no EPOLLHUP when others closed the pipe.
+        """
+        try:
+            events = self.epollObj.poll(-1)
+        except KeyboardInterrupt:
+            return False
+        except IOError as err:
+            if err.errno == errno.EINTR:
+                return False
+            else:
+                self.logErr("Error at polling fd:" + str(err))
+                raise
+
+        for fileno, event in events:
+            if fileno == self.fifo and event & select.EPOLLIN:
+                return True
+            elif fileno == self.fifo and event & (select.EPOLLERR | select.EPOLLHUP):
+                # probably some log and raise failure
+                self.logErr("exit due to unexpected events:" + str(event))
+                raise IOError(errno.EPIPE, 'hung up or error occurred while poll', self.pipePath)
+
+        return False
+                
+    def work(self):
+        """ server side daemon logic """
+        self.logInfo("server started")
+        
+        try:
+            self._openRead()
+            while True:
+                if self._waitForInput():
+                    if self._readRequest():
+                        self._executeSub()
+        except Exception as err:
+            self.logErr("Exit at error:" + str(err))
+            self._closeFIFO()
```

Comment:
do we need to `raise` here?

Reply:
Not necessarily, from the main logic, after the work ends, the process/thread will exit.